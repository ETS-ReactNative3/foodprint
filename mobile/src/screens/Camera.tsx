import React from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import {RNCamera} from 'react-native-camera';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import Axios from 'axios';
import * as qs from 'qs';

interface Props {}

const PendingView = () => (
  <View
    style={{
      flex: 1,
      backgroundColor: 'lightgreen',
      justifyContent: 'center',
      alignItems: 'center',
    }}>
    <Text>Waiting</Text>
  </View>
);

const Camera: React.FC<Props> = ({navigation}) => {
  const takePicture = async (camera: RNCamera) => {
    const options = {quality: 0.5, base64: true};
    try {
      const data = await camera.takePictureAsync(options);
      // send image to backend
      const endpoint = 'https://94a627c2.ngrok.io';
      const config = {
        headers: {'content-type': 'application/x-www-form-urlencoded'},
      };
      console.log(data);
      await Axios.post(endpoint, qs.stringify(data), config);
    } catch (err) {
      console.log('One error');
      console.log(err);
    }
  };

  return (
    <View style={styles.container}>
      <RNCamera
        captureAudio={false}
        style={styles.preview}
        type={RNCamera.Constants.Type.back}
        flashMode={RNCamera.Constants.FlashMode.off}
        androidCameraPermissionOptions={{
          title: 'Permission to use camera',
          message: 'We need your permission to use your camera to scan food',
          buttonPositive: 'Ok',
          buttonNegative: 'Cancel',
        }}
        // onGoogleVisionBarcodesDetected={({ barcodes }) => {
        //     console.log(barcodes);
        // }}
      >
        {({camera, status}) => {
          if (status !== 'READY') return <PendingView />;
          return (
            <View
              style={{flex: 0, flexDirection: 'row', justifyContent: 'center'}}>
              <TouchableOpacity
                onPress={() => takePicture(camera)}
                style={styles.capture}>
                <MaterialCommunityIcons name="food" color="grey" size={100} />
              </TouchableOpacity>
            </View>
          );
        }}
      </RNCamera>
    </View>
  );
};

export default Camera;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: 'black',
  },
  preview: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  capture: {
    overflow: 'visible',
    alignItems: 'center',
    justifyContent: 'center',
    width: 125,
    height: 125,
    borderRadius: 100,
    alignSelf: 'center',
    marginBottom: 15,
  },
});
