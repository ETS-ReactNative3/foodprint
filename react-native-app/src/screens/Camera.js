import React, { useState } from 'react';
import { TouchableOpacity, View, ActivityIndicator } from 'react-native';
import { RNCamera } from 'react-native-camera';
import { useIsFocused } from '@react-navigation/native';
import { StyleSheet } from 'react-native';

const Camera = ({ navigation }) => {
  const isFocused = useIsFocused();
  const [cameraIsReady, setCameraIsReady] = useState(true);

  const takePictureHandler = async (camera) => {
    setCameraIsReady(false);
    const options = { quality: 0.5, base64: true };
    const image = await camera.takePictureAsync(options);
    setCameraIsReady(true);
    navigation.navigate('Feedback', { file: image });
  };

  const barCodeHandler = ({ data, rawData, type, bounds }) => {
    console.log({ data, rawData, type, bounds });
    navigation.navigate('Feedback', { barcode: data });
  };

  return (
    <View accessible={true} style={styles.container}>
      {!isFocused ? <></> :
        <RNCamera
          onBarCodeRead={barCodeHandler}
          accessible={true}
          style={styles.preview}
          captureAudio={false}
          type={RNCamera.Constants.Type.back}
          flashMode={RNCamera.Constants.FlashMode.off}
          androidCameraPermissionOptions={{
            title: 'Permission to use camera',
            message: 'We need your permission to use your camera',
            buttonPositive: 'Ok',
            buttonNegative: 'Cancel',
          }}
        >
          {({ camera, status }) => {
            if (!cameraIsReady || status !== 'READY') {
              return <View style={styles.loadingContainer}>
                <ActivityIndicator
                  style={styles.noCapture} color={'white'} />
              </View>;
            }
            return (
              <View accessible={true} style={styles.nonLoadingContainer}>
                <View accessible={true} style={styles.captureContainer}>
                  <TouchableOpacity
                    accessible={true}
                    accessibilityLabel="take picture"
                    accessibilityHint="Take and upload picture to calculate carbon footprint"
                    onPress={() => takePictureHandler(camera)}
                    style={styles.capture} />
                </View>
              </View>
            );
          }}
        </RNCamera>
      }
    </View >
  );
};

export default Camera;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: 'black',
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  nonLoadingContainer: {
    flex: 1,
  },
  preview: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  captureContainer: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  capture: {
    borderWidth: 2,
    borderColor: 'white',
    alignItems: 'center',
    justifyContent: 'center',
    width: 80,
    height: 80,
    backgroundColor: 'rgba(9,9,9,0.1)',
    margin: 32,
    borderRadius: 50,
  },
  noCapture: {
    alignItems: 'center',
    justifyContent: 'center',
    width: 80,
    height: 80,
    margin: 32,
  },
});
