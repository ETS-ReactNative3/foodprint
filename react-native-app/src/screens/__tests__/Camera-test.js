import React from 'react';

import renderer from 'react-test-renderer';
import { fireEvent, render, wait } from '@testing-library/react-native';
import Camera from '../Camera';
import { useMutation } from 'react-apollo';

jest.mock('@apollo/react-hooks');

// Check current component against its snapshot
test('FoodOverview matches previous snapshot', () => {
  useMutation.mockImplementation(() => [null, { loading: null, error: null, data: null }]);
  const tree = renderer.create(<Camera />).toJSON();
  expect(tree).toMatchSnapshot();
});

// TODO mock data response from graphQL, and do more tests
test('Take picture on camera', async () => {
  // useMutation.mockImplementation(() => [null, { loading: null, error: null, data: null }]);

  const onTakePhotoEvent = jest.fn(data => data);

  const utils = render(<Camera />)
  // const snapButton = utils.getByLabelText('take picture')
  // console.log({ snapButton })
  // fireEvent.press(joinButton);
  // expect(onTakePhotoEvent.mock.calls.length).toBe(1);
});