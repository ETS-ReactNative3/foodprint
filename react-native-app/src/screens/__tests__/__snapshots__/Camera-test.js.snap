// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foodprint matches previous snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "flex": 1,
      "flexDirection": "column",
    }
  }
/>
`;
